src/relay/qnn/pass/quantize_rewrite.cc

```diff
@@ -0,0 +1,262 @@
+    const Array<IndexExpr>& out_shape) {
+  double double_multiplier = param->input_scale/param->output_scale;
+  auto scalar_multiplier = MakeConstantScalar(Float(32), double_multiplier);
+  auto input_zp = MakeConstantScalar(idtype, param->input_zero_point);
+  auto output_zp = MakeConstantScalar(Float(32), param->output_zero_point);
+
+  // Multiply the tensor with the new scale.
+  auto shifted_input_t = Subtract(input_tensor, input_zp);
+  auto casted_t = Cast(shifted_input_t, Float(32));
+  auto multiplied_t = Multiply(casted_t, scalar_multiplier);
+  auto shifted_multiplied_t = Add(output_zp, multiplied_t);
+  auto rounded_t = Round(shifted_multiplied_t);
+  auto q_imin = GetQmin(idtype);
+  auto q_imax = GetQmax(idtype);
+  auto scaled_int32_t = Cast(Clip(rounded_t, q_imin, q_imax),
+          idtype);
+
+  // Clip to the out_dtype min/max.
+  // Clip limits must be smaller than the dtype of the input tensor.
+  auto q_min = std::max(GetQmin(param->out_dtype), GetQmin(idtype));
+  auto q_max = std::min(GetQmax(param->out_dtype), GetQmax(idtype));
+  auto clipped_t = Clip(scaled_int32_t, q_min, q_max);
+  auto requantized_output = Cast(clipped_t, param->out_dtype);
+  return requantized_output;
+}
+
+/*
+ * Lowering of the requantize operation. The requantize operator converts one
+ * quantized tensor to another quantized tensor. For the output tensor, we are
+ * provided with output scale and zero point. The computation looks like this
+ *
+ * Q_output = zp_output +  (scale_input)/(scale_ouptut) * (Q_input - zp_input)
+ *
+ * The above computation can be done in floating point as the scales are in
+ * FP32. Alternatively, we can approximate floating point with fixed point
+ * computation. This is controlled by use_int_compute.
+ */
+Expr RequantizeForwardRewrite(const Call& ref_call,
+    const Array<Expr>& new_args, const NodeRef& ctx) {
+  CHECK_EQ(new_args.size(), 1);
+  Expr quantized_data = new_args[0];
+  const auto* param = ref_call->attrs.as<RequantizeAttrs>();
+
+  // Find output shape.
+  Array<IndexExpr> out_shape;
+  auto ref_call_t = ref_call->checked_type();
+  auto output_tt = ref_call_t.as<TensorTypeNode>();
+  CHECK(output_tt != nullptr) << "Type information missing."
+      << " Please run infer_type pass.";
+  out_shape = output_tt->shape;
+
+  // Find input dtype.
+  auto ref_input_t = ref_call->args[0]->checked_type();
+  auto input_tt = ref_input_t.as<TensorTypeNode>();
+  CHECK(input_tt != nullptr) << "Type information missing."
+      << " Please run infer_type pass.";
+  const auto input_dtype = input_tt->dtype;
+
+  if (param->use_int_compute) {
+    return RequantizeInt(quantized_data, param, input_dtype, out_shape);
+  } else {
+    return RequantizeFloat(quantized_data, param, input_dtype, out_shape);
+  }
+}
+
+RELAY_REGISTER_OP("qnn.requantize")
+.set_attr<FForwardRewrite>("FQuantizeForwardRewrite", RequantizeForwardRewrite);
+
+TVM_REGISTER_API("relay._qnn.rewrite")
```

Comment:
should this be ported to pass manager? @zhiics

Reply:
Talked to @anijain2305 offline, it sounds this pass is more or less a dialect that is only performed in the frontend parsers. I think we probably don't want to bring it into the pass manager because 1) it is quite frontend specific, 2) it only does `Expr` -> `Expr` transformation.

We could probably use better naming instead of `rewrite` here and `ir_pass` in Python. Since we have `pass` under the dialect fold, should we have `transform` as well? 

@tqchen thoughts?

Reply:
Yes, I think we need to use transform